### PR TITLE
Meta: Add tests using jest

### LINKED ISSUE #21: Add tests
In my latest action I extracted [the core function](https://github.com/fregante/title-to-labels-action/blob/main/parse-title.js) and [tested that](https://github.com/fregante/title-to-labels-action/blob/main/index.test.js) using the setup from the original [action template](https://github.com/actions/javascript-action).

For this action, perhaps you could extract and test the commit-to-notes part, which is the core feature:

```js
function generateReleaseNotes({commits, commitTemplate, releaseTemplate}) {
	const commitEntries = [];
	if (commits.length === 0) {
		commitEntries.push('_Maintenance release_');
	} else {
		for (const {hash, title} of commits) {
			const line = commitTemplate
				.replace('{hash}', hash)
				.replace('{title}', title)
				.replace('{url}', repoURL + '/commit/' + hash);
			commitEntries.push(line);
		}
	}
	return releaseTemplate
				.replace('{commits}', commitEntries.join('\n'))
				.replace('{range}', `[\`${range}\`](${repoURL}/compare/${range})`)
}
```

Testing the git part is harder and requires mocking, but this shouldn't be too difficult.

### PR DESCRIPTION
**Pull request by notlmn**
Closes #21

- [x] Move logic from `index.js` to `generate-release-notes.js`
- [x] Add basic tests
- [ ] ~Move more `git` commands to the other file, and mock those commands (that get tags)~
      This sounds a bit overkill to test how we get tags
- [x] Add more tests (should be fine for now, paging @fregante for more places to test)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,34 +1,25 @@
...
-  "xo": {
-    "ignore": [
-      "distribution"
-    ]
-  }
+	"private": true,
+	"scripts": {
+		"package": "ncc build index.js -o distribution",
+		"test": "xo && npm run jest",
+		"jest": "GITHUB_SERVER_URL='https://github.com' GITHUB_REPOSITORY='test/test' jest"
```
Comment (fregante) [inline package.json:9]: Since these are part of the tests, why not just specify them in the test file?

File: index.test.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
+const childProcess = require('child_process');
```
Comment (fregante) [inline index.test.js]: This file doesn't test `index.js`, so it shouldn't be called index

File: index.js [hunk 3]
```diff
@@ -2,87 +2,57 @@ const {getOctokit, context} = require('@actions/github');
...
-	} catch (error) {
-		core.setFailed(error.message);
-	}
+	core.info('Created release `' + createReleaseResponse.data.id + '` for tag `' + pushedTag + '`');
 }
 
-run();
+run().catch(error => {
+	core.setFailed(error.message);
+});
```
Comment (fregante) [inline index.js]: The PR includes some unrelated changes though, it would be easier to leave them aside (like this try/catch and the package.json cleanup)

File: index.test.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+		releaseTemplate: '{commits}',
+		exclude: '^Meta|^Lint|^Refactor'
+	});
+
+	expect(output).toEqual([
+		'- Update dependencies (#14)',
+		'- Enable `bypass-checks` on more commit statuses (#3610)',
+		'- Add tooltip to `table-input` button (#3615)',
+		'- Fix padding issue in Milestones'
+	].join('\n'));
```
Comment (fregante) [inline index.test.js]: Template literals are slightly more readable. 
```suggestion
	expect(output).toEqual(`
		- Update dependencies (#14)
		- Enable `bypass-checks` on more commit statuses (#3610)
		- Add tooltip to `table-input` button (#3615)
		- Fix padding issue in Milestones
	`);
```

To trim the whitespace you can use [common-tags](https://github.com/zspecza/common-tags#stripindent) or

```js
`


`.replace(/(\n|\b)\t+/g, '$1').trim()
```

**Review by fregante — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (notlmn) [reply package.json:9]: For some reason, setting these using `process.env` still results in them being `undefined` while getting them. Probably jest is running these tests in a subprocess?

File: index.test.js (hunk 4 above)
Comment (notlmn) [reply index.test.js]: Ended up using https://github.com/sindresorhus/strip-indent, as `common-tags` is around 222 KB and has unrelated functions as well 😄

File: package.json (hunk 1 above)
Comment (fregante) [reply package.json:9]: Set it in beforeAll https://github.com/vuejs/vue-test-utils/issues/193#issuecomment-528358591

Comment (notlmn) [reply package.json:9]: That still doesn't seem to fix it, tried `beforeAll` as well.

image

Comment (notlmn) [reply package.json:9]: Got it, because envs have to be set before `require('./generate-release-notes')`, as `process.env` is accessed at the top of that module.

Comment (fregante) [reply package.json:9]: Indeed I usually solve that by importing a fixtures.js file, so it’s executed before other imports
